### PR TITLE
[doc] Revise the CLI documentation

### PR DESCRIPTION
**Pull request by tengqm**
This PR updates the CLI documentation.
The revised version retains all existing information, and the major
revision is about simplification and fixing ambiguity or typo errors.

### CODE REVIEW

**Comment by tengqm** [comment]
To reviewers: checking the git diff is not an efficient way to review for this PR (drastical change). Maybe a direct checking of the rendered output can help save some energy.

File: docs/cli.md [hunk 1]
```diff
@@ -8,25 +8,29 @@ last_update:
 license: 'This software is licensed under the Apache License version 2.'
 ---
 
-This document provides guidance on managing metadata within Apache Gravitino using the Command Line Interface (CLI). The CLI offers a terminal based alternative to using code or the REST interface for metadata management.
+This document provides a guidance on managing metadata within Apache Gravitino using
+the Command Line Interface (CLI). The CLI offers a terminal based alternative to
+using code or the REST API for metadata management.
```
Comment (justinmclean) [inline docs/cli.md:13]: Please don't break up sentences, this stops some IDE tools like grammar checkers from working and makes it easier for errors to slip into the text.

File: docs/cli.md [hunk 2]
```diff
@@ -8,25 +8,29 @@ last_update:
...
 
-Currently, the CLI allows users to view metadata information for metalakes, catalogs, schemas, tables, users, roles, groups, tags, topics and filesets. Future updates will expand on these capabilities.
+Currently, the CLI allows users to view metadata information for metalakes, catalogs,
+schemas, tables, users, roles, groups, tags, topics and filesets.
+Future updates will expand on these capabilities.
 
 ## Running the CLI
 
-You can configure an alias for the CLI for ease of use, with the following command:
+You can configure an alias for the CLI for ease of use. For example:
```
Comment (justinmclean) [inline docs/cli.md:21]: This is not an example.

File: docs/cli.md [hunk 3]
```diff
@@ -64,54 +68,60 @@ The general structure for running commands with the Gravitino CLI is `gcli entit
...
-- remove: Remove a property from an entity
-- properties: Display an entities properties
+- `create`: Create a new entity
+- `delete`: Delete an existing entity
+- `details`: Show detailed information about an entity
+- `list`: List available entities
+- `properties`: Display an entity's properties
+- `remove`: Remove a property from an entity
+- `set`: Set a property on an entity
+- `update`: Update an existing entity
```
Comment (justinmclean) [inline docs/cli.md:84]: Please leave the order as it, it ordered according to use of command / logical grouping. The list and details commands belong together, so do create/delete/update and set/remove. Alphabetical order is not helpful here.

File: docs/cli.md [hunk 4]
```diff
@@ -64,54 +68,60 @@ The general structure for running commands with the Gravitino CLI is `gcli entit
...
-As dealing with one Metalake is a typical scenario, you can set the Metalake name in several ways so it doesn't need to be passed on the command line.
+Considering that managing a metalake is a typical scenario, you can set the metalake
+name in several ways so it doesn't need to be passed on the command line.
 
-1. Passed in on the command line via the `--metalake` parameter.
-2. Set via the `GRAVITINO_METALAKE` environment variable.
-3. Stored in the Gravitino CLI configuration file.
+1. Specify the `--metalake` argument on the command line.
+1. Set the `GRAVITINO_METALAKE` environment variable.
+1. Save it in the Gravitino [CLI configuration file](#config-file).
```
Comment (justinmclean) [inline docs/cli.md:93]: This change (Save it) doesn't work well.

File: docs/cli.md [hunk 5]
```diff
@@ -64,54 +68,60 @@ The general structure for running commands with the Gravitino CLI is `gcli entit
  -x,--index              display index information
  -z,--provider <arg>     provider one of hadoop, hive, mysql, postgres,
                          iceberg, kafka
- ```
+```
 
-## Commands
+## Commands Types
```
Comment (justinmclean) [inline docs/cli.md:73]: I would not use "Types" to describe commands. Perhaps "Command Actions"?

File: docs/cli.md [hunk 6]
```diff
@@ -8,25 +8,29 @@ last_update:
 license: 'This software is licensed under the Apache License version 2.'
 ---
 
-This document provides guidance on managing metadata within Apache Gravitino using the Command Line Interface (CLI). The CLI offers a terminal based alternative to using code or the REST interface for metadata management.
+This document provides a guidance on managing metadata within Apache Gravitino using
```
Comment (justinmclean) [inline docs/cli.md:11]: "a guidance" -> "guidance" as the guidance in the document is not singular

File: docs/cli.md [hunk 7]
```diff
@@ -64,54 +68,60 @@ The general structure for running commands with the Gravitino CLI is `gcli entit
...
 
-The command line option overrides the environment variable and the environment variable overrides the configuration file.
+The command line option overrides the environment variable and the environment variable
+takes precedence over the configuration file.
 
 ### Setting the Gravitino URL
 
-As you need to set the Gravitino URL for every command, you can set the URL in several ways.
+To avoid specifying the Gravitino URL using `--url` argument for every command,
+you can persist the the URL setting.
```
Comment (justinmclean) [inline docs/cli.md:101]: "the the URL setting." -> " the URL setting."

File: docs/cli.md [hunk 8]
```diff
@@ -152,710 +163,464 @@ keytabFile=file.keytab
...
+`gcli catalog create --provider hive --name <name> --properties <property key-value pairs>`
+: Create a catalog.
 
-```bash
-gcli catalog create --name hive --provider hive --properties metastore.uris=thrift://hive-host:9083
-```
+Different catalogs require different properties.
+For example, a Hive catalog requires a `--metastore-uris` property.
+ 
+- For a Hive cata log, you need to specify the `metastore.uris` property. For example:
```
Comment (justinmclean) [inline docs/cli.md:278]: "cata log," -> "catalog,"

File: docs/cli.md [hunk 9]
```diff
@@ -152,710 +163,464 @@ keytabFile=file.keytab
...
-gcli tag update --tag tagA --comment "new comment"
-```
+`gcli tag update --tag <tag> --comment "<new-comment>"`
+: Update the comment for a tag.
 
 ### Owner commands
 
-#### List an owner
+`gcli catalog details --name <catalog> --owner`
+: Show the details about an owner for a catalog.
```
Comment (justinmclean) [inline docs/cli.md:492]: "an owner for a catalog." - > "the owner of a catalog."

File: docs/cli.md [hunk 10]
```diff
@@ -152,710 +163,464 @@ keytabFile=file.keytab
...
 
-```bash
-gcli catalog details --owner --name postgres
-```
+`gcli catalog set --name <catalog> --owner --user <user>`
+: Set the specified user as the owner for the specified catalog.
```
Comment (justinmclean) [inline docs/cli.md:495]: "or the specified catalog." -> "of the specified catalog."

File: docs/cli.md [hunk 11]
```diff
@@ -152,710 +163,464 @@ keytabFile=file.keytab
...
-gcli catalog set --owner --user admin --name postgres
-```
-
-#### Set an owner to a group
-
-```bash
-gcli catalog set --owner --group groupA --name postgres
-```
+`gcli catalog set --name <catalog> --owner --group <group>`
+: Set the specified group as the owner for the catalog.
```
Comment (justinmclean) [inline docs/cli.md:498]: "for the catalog." -> "of the catalog." also add or remove specified so it is consistent with the above.

File: docs/cli.md [hunk 12]
```diff
@@ -152,710 +163,464 @@ keytabFile=file.keytab
...
 
-```bash
-gcli column create --name catalog_postgres.hr.departments.value --datatype long
-gcli column create --name catalog_postgres.hr.departments.money --datatype "decimal(10,2)"
-gcli column create --name catalog_postgres.hr.departments.name --datatype "varchar(100)"
-gcli column create --name catalog_postgres.hr.departments.fullname --datatype "varchar(250)" --default "Fred Smith" --null=false
-```
+`gcli column create --name <column> --datatype long`
+: Create a column. You can specify the default value using the `--default` property,
+  and you can specify whether the column can be null using the `--null` option.
```
Comment (justinmclean) [inline docs/cli.md:602]: Having several examples here would be helpful to a user.

File: docs/cli.md (hunk 1 above)
Comment (tengqm) [reply docs/cli.md:13]: Em... Let me try articulate the reasons behind this kind of changes.

- We'd better treat docs as code. That means we write them and then track whatever
  changes in the future.
- The GIT tool compare changes on a line by line basis. When we have a long line of
  text, GIT can only tell you that this "line" has changed. We as human reviewers have
  to identify the actual change(s) manually. That means a painful experience like finding
  a needle fallen on the ground.
- The GitHub web UI works the same way. I have myself been tortured by this for many years.
- When we break long lines manually, appropriately, this line breaks are not impacting
  the rendered output (mainly HTML here).
- With manual line breaks, we ensure that every developer sees the same thing on his
  or her environment. Think about the tab size problem here.
- I am not quite sure how an IDE helps improve the documentation. Even if there is an
  IDE (say IDEA) can do a great job for something, we are not writing code/docs for
  that IDE. Human is not supposed to be hijacked by tools, right?
- I'm open to learn the draw backs of manually wrapped long lines. In some other
  communities, manually wrapping long lines is strongly encouraged.
- This does NOT mean we want to encourage people to break the long lines at a fix boundary.
  The recommended way is to break them WHEN APPROPRIATE. For example, we encourage
  breaking long lines when a statement ends, when there is a punctuation character,
  so on and so forth. We discourage breaking links, however, because it doesn't make
  a lot of senses.

File: docs/cli.md (hunk 3 above)
Comment (tengqm) [reply docs/cli.md:84]: As I have responded in another review comment. When we don't have a consensus on the
ordering of things, the best way to do it is to sort them in alphabetic order.
Why? Every people may have a different opinion on the definition of "logical grouping".
For example, why the first entry should be `list`? What am I doing this if I haven't
created something? Why `delete` comes before `update`?
The main problem of this "logical grouping" is about predictability. There will never
be a consensus on the correct/golden ordering.
With that, I'm strongly against it.

File: docs/cli.md (hunk 1 above)
Comment (justinmclean) [reply docs/cli.md:13]: The GitHub web UI doesn't work the same way it highlights word changes within a line.

We are writing docs in an IDE, and IDE plugins such as grammar checkers help do that. They don't always work if you break sentences over lines.

In this project, I noticed that in the documentation that breaks up sentences,  errors are easily introduced in changes (as they are here).

File: docs/cli.md (hunk 6 above)
Comment (tengqm) [reply docs/cli.md:11]: OK.

File: docs/cli.md (hunk 2 above)
Comment (tengqm) [reply docs/cli.md:21]: Well .. it is. It is using a "incubating", "SNAPSHOT" jar.
If the syntax is changed to `alias gcli='java -jar <path-to-jar>`, I would not call it an example.

File: docs/cli.md (hunk 5 above)
Comment (tengqm) [reply docs/cli.md:73]: Revised.

File: docs/cli.md (hunk 4 above)
Comment (tengqm) [reply docs/cli.md:93]: Yup. Good suggestion.

File: docs/cli.md (hunk 7 above)
Comment (tengqm) [reply docs/cli.md:101]: Fixed.

File: docs/cli.md (hunk 8 above)
Comment (tengqm) [reply docs/cli.md:278]: Good spot.

File: docs/cli.md [hunk 13]
```diff
@@ -152,710 +163,464 @@ keytabFile=file.keytab
...
 
 ### Table commands
 
-When creating a table the columns are specified in CSV file specifying the name of the column, the datatype, a comment, true or false if the column is nullable, true or false if the column is auto incremented, a default value and a default type. Not all of the columns need to be specifed just the name and datatype columns. If not specified comment default to null, nullability to true and auto increment to false. If only the default value is specified it defaults to the same data type as the column.
+When creating a table, the columns are specified in a CSV file containing the properties
+of each column, including
+
+- the column name (required)
+- the data type (required)
+- the comment string which defaults to null
```
[the remark this answers is not in the document]
Comment (tengqm) [reply docs/cli.md:366]: ok

File: docs/cli.md [hunk 14]
```diff
@@ -152,710 +163,464 @@ keytabFile=file.keytab
...
+- a string format boolean indicating whether the column can be null (default=true)
+- a string format boolean indicating whether the column is auto-incremented (default=false)
+- the default value
+- the default data type, defaults to the column data type if a default value is specified.
```
[the remark this answers is not in the document]
Comment (tengqm) [reply docs/cli.md:370]: I think we'd better mention that, because we are not saying
"if a default type is specified".
The defaulting of "type" depends on the setting of "value".
We need to highlight this.

File: docs/cli.md (hunk 9 above)
Comment (tengqm) [reply docs/cli.md:492]: Correct. Got it.

File: docs/cli.md (hunk 10 above)
Comment (tengqm) [reply docs/cli.md:495]: done.

File: docs/cli.md (hunk 11 above)
Comment (tengqm) [reply docs/cli.md:498]: Agreed.

File: docs/cli.md (hunk 12 above)
Comment (tengqm) [reply docs/cli.md:602]: OK

File: docs/cli.md (hunk 1 above)
Comment (tengqm) [reply docs/cli.md:13]: image

Let's use the change I made as an example. GitHub failed to identify that this change is only about inserting several line breaks.

File: docs/cli.md (hunk 3 above)
Comment (justinmclean) [reply docs/cli.md:84]: While there may be other logical ways to order them, that doesn't mean we should do it alphabetically. `list` is first as it is most likely the first command a user will want to use, followed by `details` the next most likely command a user would use and `list` and `details` are commonly used together. `delete` pairs with `create` so that is why it comes before `update`. If the commands were slightly different, I'd would have gone with a more CRUD style ordering.

As a native English speaker, I would not expect this to be in alphabetical order. Alphabetical order is used when you have a long list. If there were, say 20 options, I would sort them alphabetically or break them up into groups. Not in the `--help` command is sorted alphabetically, but I had considered following this advice from https://clig.dev/#help 

"Display the most common flags and commands at the start of the help text. It’s fine to have lots of flags, but if you’ve got some really common ones, display them first. For example, the Git command displays the commands for getting started and the most commonly used subcommands first:"

There is also actual usability research on this:
https://www.nngroup.com/articles/alphabetical-sorting-must-mostly-die/

Basically, when users don't know the name of the thing they want, it makes A–Z listings useless and A–Z listings are directly harmful because they hide logical ordering.

File: docs/cli.md (hunk 1 above)
Comment (justinmclean) [reply docs/cli.md:13]: That shows no changes other than the newlines or lines moved. This is a better example:
image

Comment (justinmclean) [reply docs/cli.md:13]: Here you can see two words have been changed

File: docs/cli.md (hunk 2 above)
Comment (justinmclean) [reply docs/cli.md:21]: Because if you compile it, that is what you always get, but it's a minor thing. This change probably doesn't matter in teh scheme of things.

File: docs/cli.md (hunk 3 above)
Comment (tengqm) [reply docs/cli.md:84]: @justinmclean That is persuasive. I was most of the time treating this docs as reference
rather than a tutorial. In a tutorial or guideline, we leave the least frequently used
(aka, "advanced") commands at the bottom of the list.
In this case, maybe, as you said, we can keep the short list as is.

File: docs/cli.md (hunk 2 above)
Comment (tengqm) [reply docs/cli.md:21]: Okay. I'm not strongly opinionated at this.

Comment (justinmclean) [reply docs/cli.md:21]: Neither am I

File: docs/cli.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (justinmclean) [reply docs/cli.md:370]: You need to specify the column data type and optionally a default value. If you specify the value you can optionally set the data type of that value. If you do not specify the value's default type, it will use the column's data type. They are likely to be the same in 99% of cases, I'd guess.

[the remark this answers is not in the document]
Comment (tengqm) [reply docs/cli.md:370]: Let me try rephrase this ... you mean the following is true no matter a default value is specified or not, right?

> the default data type, defaults to the column data type

[the remark this answers is not in the document]
Comment (justinmclean) [reply docs/cli.md:370]: No as if you don't specify a default value, there is no default value to have a default type.  If you do specify a default value but not a default type it will default to the column's data type, OR you can specify a default value and a default type.

[the remark this answers is not in the document]
Comment (tengqm) [reply docs/cli.md:370]: Thanks for the clarification. Then I believe the if-clause is required.

[the remark this answers is not in the document]
Comment (justinmclean) [reply docs/cli.md:370]: The above text "the default data type, defaults to the column data type if a default value is specified." is incorrect because if you specify a default value and data type, it will not default to the column data type.

[the remark this answers is not in the document]
Comment (justinmclean) [reply docs/cli.md:370]: Also (very minor), if you end in a full stop, you probably want to start with a capital and do the same for the other points.

[the remark this answers is not in the document]
Comment (justinmclean) [reply docs/cli.md:370]: Some people would take that sentence as: If you specify a default value, then the default column type defaults to the column data type, whether you specified that or not.

[the remark this answers is not in the document]
Comment (tengqm) [reply docs/cli.md:370]: > If you specify a default value, then the default column type defaults to the column data type

Yes. That is exactly what I captured from your clarification.
Maybe you can provide a clearer description?

BTW, this logic is ugly and needs some rethinking. Are we sure we want to do this?
You can see I have had a difficult time to understand this logic.
The same difficulty will be a pain for other users/developers.
If we have a logic that is so difficult to explain, I'd suggest we drop this smart hack.

[the remark this answers is not in the document]
Comment (justinmclean) [reply docs/cli.md:370]: I'll come up with some better text. In 99% of cases, the column will be omitted, I think most people will just follow the example and not even notice.

**Comment by tengqm** [comment]
superceded by #7111
